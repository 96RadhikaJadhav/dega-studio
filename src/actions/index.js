import { GET_ORGANISATIONS, GET_ORGANISATIONS_SUCCESS } from './types';
import axios from 'axios';

export const getOrganisations = () => {
  return (dispatch, getState) => {};
};

const getOrganisationsSuccess = (todo) => ({
  type: GET_ORGANISATIONS_SUCCESS,
  payload: {},
});
