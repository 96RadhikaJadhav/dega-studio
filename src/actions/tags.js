import axios from '../utils/axios';
import {
  GET_TAGS_SUCCESS,
  GET_TAGS_FAILURE,
  ADD_TAG_FAILURE,
  ADD_TAG_SUCCESS,
  API_ADD_TAG,
  API_GET_TAGS,
  UPDATE_TAG_FAILURE,
  UPDATE_TAG_SUCCESS,
  DELETE_TAG_SUCCESS,
  DELETE_TAG_FAILURE,
  LOADING_TAGS,
} from '../constants/tags';

export const getTags = (query) => {
  return async (dispatch, getState) => {
    let found = false;
    const {
      tags: { req },
    } = getState();

    // map data based on query
    req.forEach((each) => {
      const { limit, page } = each.query;
      if (page === query.page && limit === query.limit) {
        found = true;
        return;
      }
    });

    if (!found) {
      dispatch(loadingSpaces());
      const response = await axios({
        url: API_GET_TAGS,
        method: 'get',
        params: query,
      }).catch((error) => {
        dispatch(getTagsFailure(error.message));
      });
      if (response) {
        dispatch(getTagsSuccess(response.data, query));
      }
    }
  };
};

export const addTag = (data) => {
  return async (dispatch, getState) => {
    dispatch(loadingSpaces());
    const response = await axios({
      url: API_ADD_TAG,
      method: 'post',
      data: { ...data },
    }).catch((error) => {
      dispatch(addTagFailure(error.message));
    });
    if (response) {
      dispatch(addTagSuccess(data));
    }
  };
};

export const updateTag = (data) => {
  return async (dispatch, getState) => {
    dispatch(loadingSpaces());

    const response = await axios({
      url: API_ADD_TAG + `/${data.id}`,
      method: 'put',
      data: { ...data },
    }).catch((error) => {
      dispatch(updateTagFailure(error.message));
    });
    if (response) {
      dispatch(updateTagSuccess(data));
    }
  };
};

export const deleteTag = (id) => {
  return async (dispatch, getState) => {
    dispatch(loadingSpaces());

    const response = await axios({
      url: API_ADD_TAG + `/${id}`,
      method: 'delete',
    }).catch((error) => {
      dispatch(deleteTagFailure(error.message));
    });
    if (response) {
      dispatch(deleteTagSuccess(id));
    }
  };
};

const loadingSpaces = () => ({
  type: LOADING_TAGS,
});

const getTagsSuccess = (data, query) => ({
  type: GET_TAGS_SUCCESS,
  payload: { data, query },
});

const getTagsFailure = (error) => ({
  type: GET_TAGS_FAILURE,
  payload: {
    error,
  },
});

const addTagSuccess = (data) => ({
  type: ADD_TAG_SUCCESS,
  payload: {
    ...data,
  },
});

const addTagFailure = (error) => ({
  type: ADD_TAG_FAILURE,
  payload: {
    error,
  },
});

const updateTagSuccess = (data) => ({
  type: UPDATE_TAG_SUCCESS,
  payload: {
    ...data,
  },
});

const updateTagFailure = (error) => ({
  type: UPDATE_TAG_FAILURE,
  payload: {
    error,
  },
});

const deleteTagSuccess = (id) => ({
  type: DELETE_TAG_SUCCESS,
  payload: id,
});

const deleteTagFailure = (error) => ({
  type: DELETE_TAG_FAILURE,
  payload: {
    error,
  },
});
