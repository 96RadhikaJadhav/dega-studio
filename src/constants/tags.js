//Actions
export const GET_TAGS_SUCCESS = 'GET_TAGS_SUCCESS';
export const GET_TAGS_FAILURE = 'GET_TAGS_FAILURE';
export const ADD_TAG_SUCCESS = 'ADD_TAG_SUCCESS';
export const ADD_TAG_FAILURE = 'ADD_TAG_FAILURE';
export const UPDATE_TAG_SUCCESS = 'UPDATE_TAG_SUCCESS';
export const UPDATE_TAG_FAILURE = 'UPDATE_TAG_FAILURE';
export const DELETE_TAG_SUCCESS = 'DELETE_TAG_SUCCESS';
export const DELETE_TAG_FAILURE = 'DELETE_TAG_FAILURE';
export const LOADING_TAGS = 'LOADING_TAGS';

//API
export const API_GET_TAGS = '/core/tags';
export const API_ADD_TAG = '/core/tags';

//Notification
