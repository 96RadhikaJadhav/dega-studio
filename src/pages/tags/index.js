import React from 'react';
import TagsList from './components/TagsList';
import { Space, Button } from 'antd';
import { Link } from 'react-router-dom';

function Tags() {
  return (
    <Space direction="vertical">
      <Link key="1" to="/tags/create">
        <Button>Create New</Button>
      </Link>
      <TagsList />
    </Space>
  );
}

export default Tags;
