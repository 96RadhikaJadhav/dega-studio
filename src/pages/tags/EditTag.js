import React from 'react';
import TagsCreateForm from './components/TagsCreateForm';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';
import { updateTag, getTag } from '../../actions/tags';
import { useHistory } from 'react-router-dom';
import { useParams } from 'react-router-dom';

function EditTag() {
  const history = useHistory();
  const { id } = useParams();

  const dispatch = useDispatch();
  const { tag, loading } = useSelector((state) => {
    return {
      tag: state.tags.details[id] ? state.tags.details[id] : null,
      loading: state.tags.loading,
    };
  });

  React.useEffect(() => {
    dispatch(getTag(id));
  }, [dispatch, id]);

  if (loading) return <Skeleton />;

  const onUpdate = (values) => {
    dispatch(updateTag({ ...tag, ...values }));
    history.push('/tags');
  };
  return <TagsCreateForm data={tag} onCreate={onUpdate} />;
}

export default EditTag;
