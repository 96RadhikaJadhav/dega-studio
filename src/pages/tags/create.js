import React from 'react';
import { useDispatch } from 'react-redux';
import { addTag } from '../../actions/tags';
import { useHistory } from 'react-router-dom';
import TagCreateForm from './components/TagsCreateForm';

function CreateTag() {
  const history = useHistory();

  const dispatch = useDispatch();
  const onCreate = (values) => {
    dispatch(addTag(values));
    history.push('/tags');
    console.log('Received values of form: ', values);
  };
  return <TagCreateForm onCreate={onCreate} />;
}

export default CreateTag;
