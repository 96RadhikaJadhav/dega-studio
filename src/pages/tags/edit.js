import React from 'react';
import TagsCreateForm from './components/TagsCreateForm';
import { useDispatch, useSelector } from 'react-redux';
import { addTag } from '../../actions/tags';
import { useHistory } from 'react-router-dom';
import useQuery from '../../utils/useQuery';
import _ from 'lodash';

function EditTag() {
  const history = useHistory();
  const query = useQuery();
  const id = query.get('id');

  const dispatch = useDispatch();
  const { tag } = useSelector((state) => {
    const { tags, ...other } = state.tags;
    const tag = _.find(tags, { id: parseInt(id) });
    return {
      tag,
      ...other,
    };
  });

  const onCreate = (values) => {
    dispatch(addTag(values));
    history.push('/tags');
    console.log('Received values of form: ', values);
  };
  return <TagsCreateForm data={tag} onCreate={onCreate} />;
}

export default EditTag;
