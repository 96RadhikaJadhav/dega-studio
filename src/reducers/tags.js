import { GET_TAGS_SUCCESS, ADD_TAG_SUCCESS } from '../constants/tags';
import { LOADING_PAGE } from '../constants';

const initialState = {
  tags: [],
  loading: false,
};

export default function tagsReducer(state = initialState, action = {}) {
  if (!action.payload) {
    return state;
  }
  console.log('tags Reducer', state, action);
  switch (action.type) {
    case LOADING_PAGE:
      return {
        ...state,
        loading: true,
      };
    case GET_TAGS_SUCCESS:
      return {
        ...state,
        tags: action.payload.nodes,
        loading: false,
      };
    case ADD_TAG_SUCCESS:
      return {
        ...state,
        tags: [...state.tags, action.payload],
        loading: false,
      };
    default:
      return state;
  }
}
