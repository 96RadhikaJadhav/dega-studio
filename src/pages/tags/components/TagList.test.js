import React from 'react';
import { BrowserRouter as Router, Link } from 'react-router-dom';
import renderer, { act } from 'react-test-renderer';
import { useSelector, useDispatch } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { Popconfirm, Button, Table } from 'antd';

import '../../../matchMedia.mock';
import TagList from './TagList';
import { getTags, deleteTag } from '../../../actions/tags';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));
jest.mock('../../../actions/tags', () => ({
  getTags: jest.fn(),
  deleteTag: jest.fn(),
}));

describe('Tags List component', () => {
  let store;
  let mockedDispatch;
  const tag = {
    id: 1,
    name: 'tag',
    slug: 'slug',
    description: 'description',
  };

  describe('snapshot testing', () => {
    beforeEach(() => {
      store = mockStore({});
      store.dispatch = jest.fn();
      mockedDispatch = jest.fn();
      useDispatch.mockReturnValue(mockedDispatch);
    });
    it('should render the component', () => {
      useSelector.mockImplementation((state) => ({}));
      const tree = renderer.create(<TagList />).toJSON();
      expect(tree).toMatchSnapshot();
      expect(useSelector).toHaveBeenCalled();
    });
    it('should match component when loading', () => {
      useSelector.mockImplementation((state) => ({
        tags: [],
        total: 0,
        loading: true,
      }));
      const tree = renderer.create(<TagList />).toJSON();
      expect(tree).toMatchSnapshot();
      expect(useSelector).toHaveBeenCalled();
    });
    it('should match component with tags', () => {
      useSelector.mockImplementation((state) => ({
        tags: [tag],
        total: 1,
        loading: false,
      }));

      let component;
      act(() => {
        component = renderer.create(
          <Router>
            <TagList />
          </Router>,
        );
      });
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      expect(useSelector).toHaveBeenCalled();
      expect(mockedDispatch).toHaveBeenCalledTimes(1);
      expect(useSelector).toHaveReturnedWith({
        tags: [tag],
        total: 1,
        loading: false,
      });
      expect(getTags).toHaveBeenCalledWith({ page: 1 });
    });
  });
  describe('component testing', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      mockedDispatch = jest.fn(() => new Promise((resolve) => resolve(true)));
      useDispatch.mockReturnValue(mockedDispatch);
    });
    it('should change the page', () => {
      useSelector.mockImplementation((state) => ({}));

      const wrapper = shallow(<TagList />);
      const table = wrapper.find(Table);
      table.props().pagination.onChange(2);
      wrapper.update();
      const updatedTable = wrapper.find(Table);
      expect(updatedTable.props().pagination.current).toEqual(2);
    });
    it('should delete tag', () => {
      useSelector.mockImplementation((state) => ({
        tags: [tag],
        total: 1,
        loading: false,
      }));

      const wrapper = mount(
        <Router>
          <TagList />
        </Router>,
      );
      const button = wrapper.find(Button).at(1);
      expect(button.text()).toEqual('Delete');

      button.simulate('click');
      const popconfirm = wrapper.find(Popconfirm);
      popconfirm
        .findWhere((item) => item.type() === 'button' && item.text() === 'OK')
        .simulate('click');

      expect(deleteTag).toHaveBeenCalled();
      expect(deleteTag).toHaveBeenCalledWith(1);
      expect(getTags).toHaveBeenCalledWith({ page: 1 });
    });
    it('should edit tag', () => {
      useSelector.mockImplementation((state) => ({
        tags: [tag],
        total: 1,
        loading: false,
      }));

      const wrapper = mount(
        <Router>
          <TagList />
        </Router>,
      );
      const link = wrapper.find(Link).at(0);
      const button = link.find(Button).at(0);
      expect(button.text()).toEqual('Edit');
      expect(link.prop('to')).toEqual('/tags/1/edit');
    });
    it('should have no delete and edit buttons', () => {
      useSelector.mockImplementation((state) => ({}));

      const wrapper = mount(
        <Router>
          <TagList />
        </Router>,
      );

      const button = wrapper.find(Button);
      expect(button.length).toEqual(0);
    });
  });
});
