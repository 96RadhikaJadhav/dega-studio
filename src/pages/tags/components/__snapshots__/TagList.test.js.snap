// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags List component snapshot testing should match component when loading 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <TagList />
  </BrowserRouter>
</ContextProvider>
`;

exports[`Tags List component snapshot testing should match component with tags 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <TagList />
  </BrowserRouter>
</ContextProvider>
`;

exports[`Tags List component snapshot testing should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <TagList />
  </BrowserRouter>
</ContextProvider>
`;
